pkg/dependencies/factory.go

```diff
@@ -219,29 +202,112 @@ func (f *Factory) WithProviderFactory(clusterConfigFile string, clusterConfig *v
 	case v1alpha1.DockerDatacenterKind:
 		f.WithDocker().WithKubectl()
 	case v1alpha1.TinkerbellDatacenterKind:
-		f.WithKubectl().WithTink(clusterConfigFile).WithPbnj(clusterConfigFile)
+		f.WithKubectl().WithTink(clusterConfigFile).WithPbnj(clusterConfigFile).WithWriter()
 	case v1alpha1.SnowDatacenterKind:
 		f.WithKubectl().WithSnowConfigManager()
 	}
 
 	f.buildSteps = append(f.buildSteps, func(ctx context.Context) error {
-		if f.providerFactory != nil {
+		if f.dependencies.Provider != nil {
 			return nil
 		}
 
-		f.providerFactory = &factory.ProviderFactory{
-			DockerClient:              f.dependencies.DockerClient,
-			DockerKubectlClient:       f.dependencies.Kubectl,
-			CloudStackCmkClient:       f.dependencies.Cmk,
-			CloudStackKubectlClient:   f.dependencies.Kubectl,
-			VSphereGovcClient:         f.dependencies.Govc,
-			VSphereKubectlClient:      f.dependencies.Kubectl,
-			SnowKubectlClient:         f.dependencies.Kubectl,
-			SnowConfigManager:         f.dependencies.SnowConfigManager,
-			TinkerbellKubectlClient:   f.dependencies.Kubectl,
-			TinkerbellClients:         tinkerbell.TinkerbellClients{ProviderTinkClient: f.dependencies.Tink, ProviderPbnjClient: f.dependencies.Pbnj},
-			Writer:                    f.dependencies.Writer,
-			ClusterResourceSetManager: f.dependencies.ResourceSetManager,
+		switch clusterConfig.Spec.DatacenterRef.Kind {
```

Comment:
nit: we could avoid a second switch by encapsulating the building logic in a closure (for example) in the first switch